astropy/timeseries/periodograms/lombscargle/implementations/utils.py

```diff
@@ -79,7 +102,273 @@ def extirpolate(x, y, N=None, M=4):
     return result
 
 
-def trig_sum(t, h, df, N, f0=0, freq_factor=1, oversampling=5, use_fft=True, Mfft=4):
+def jn(n, x, num_terms=10):
+    """
+    Approximate the Bessel function of the first kind
+    of an integer order n for small values of x.
+    Approximately equivalent to scipy.special.jv(n, x)
+    """
+    n = np.asarray(n, dtype=np.int32)
+    x = np.asarray(x)
+    half_x = 0.5 * x
+
+    n = n.astype(np.int64)
+
+    # Keep the array longer to approximately ignore the factorials of negative integers in the denominator
+    n_max = int(np.max(n)) + num_terms + 10
+    fact = np.empty(n_max + 1, dtype=np.float64)
+    fact[0] = 1.0
+    for i in range(1, n_max + 1):
+        fact[i] = fact[i - 1] * i
+
+    result = np.zeros(np.broadcast(n, x).shape, dtype=np.float64)
+
+    half_x2 = half_x * half_x
+    term_numer = half_x**n
+    term_denom = fact[n]
+
+    for k in range(num_terms):
+        term_denom = fact[k] * fact[(n + k)]
+        result += term_numer / term_denom
+        term_numer *= -half_x2
+
+    return result
+
+
+def get_lra_params(x, N, eps):
+    """
+    Compute parameters for the Low Rank Approximation (LRA) NUFFT.
+
+    Parameters
+    ----------
+    x : array-like
+        Non-uniform sample locations, shape (n,).
+    N : int
+        Required number of grid points in the frequency domain.
+    eps : float
+        Desired relative error tolerance.
+
+    Returns
+    -------
+    s : ndarray, int
+        Wrapped grid indices, shape (n,).
+    gamma : float
+        Maximum absolute deviation from nearest uniform grid point.
+    K : int
+        Number of FFTs required to achieve desired precision (eps) when round-off errors are ignored.
+    """
+    s = np.mod(np.rint(N * x), N).astype(int)
+    er = (N * x - np.rint(N * x) + 0.5) % 1 - 0.5
+    gamma = np.max(np.abs(er))
+    if gamma <= eps:
+        K = 1
+    if gamma <= eps:
+        K = 1
+    else:
+        lut = {
+            2: 2.6e-1,
+            3: 5.4e-2,
+            4: 8.6e-3,
+            5: 1.1e-3,
+            6: 1.3e-4,
+            7: 1.2e-5,
+            8: 1.2e-6,
+            9: 9.0e-8,
+            10: 6.8e-9,
+            11: 4.3e-10,
+            12: 2.9e-11,
+            13: 1.6e-12,
+            14: 8.4e-14,
+            15: 2.2e-14,
+            16: 4.6e-16,
+        }
+
+        # Find the smallest K in the LUT where eps is less than or equal to the LUT value
+        K = 16  # Default to K=16 if eps is too small
```

Comment:
this assignment could be move to an `else` branch on the `for` loop
https://docs.python.org/3/tutorial/controlflow.html#else-clauses-on-loops